pkg/openstack-cinder/config/configsync.go

```diff
@@ -0,0 +1,220 @@
+	"k8s.io/klog/v2"
+)
+
+// This ConfigSyncController translates the ConfigMap provided by the user
+// containing configuration information for the Cinder CSI driver.
+type ConfigSyncController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	configMapLister      corelisters.ConfigMapLister
+	infrastructureLister configv1listers.InfrastructureLister
+	eventRecorder        events.Recorder
+}
+
+const (
+	sourceConfigKey   = "config"
+	targetConfigKey   = "cloud.conf"
+	enableTopologyKey = "enable_topology"
+
+	infrastructureResourceName = "cluster"
+)
+
+func NewConfigSyncController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	configInformers configinformers.SharedInformerFactory,
+	resyncInterval time.Duration,
+	eventRecorder events.Recorder) factory.Controller {
+
+	// Read configmap from user-managed namespace and save the translated one
+	// to the operator namespace
+	configMapInformer := informers.InformersFor(util.OpenShiftConfigNamespace)
+	c := &ConfigSyncController{
+		operatorClient:       operatorClient,
+		kubeClient:           kubeClient,
+		configMapLister:      configMapInformer.Core().V1().ConfigMaps().Lister(),
+		infrastructureLister: configInformers.Config().V1().Infrastructures().Lister(),
+		eventRecorder:        eventRecorder.WithComponentSuffix("ConfigSync"),
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(resyncInterval).WithSyncDegradedOnError(operatorClient).WithInformers(
+		operatorClient.Informer(),
+		configMapInformer.Core().V1().ConfigMaps().Informer(),
+	).ToController("ConfigSync", eventRecorder)
+}
+
+func (c *ConfigSyncController) Name() string {
+	return "ConfigSyncController"
+}
+
+func (c *ConfigSyncController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	var err error
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	enableTopologyFeature, err := enableTopologyFeature()
+	if err != nil {
+		return err
+	}
+
+	infra, err := c.infrastructureLister.Get(infrastructureResourceName)
+	if err != nil {
+		return err
+	}
+
+	var sourceConfig *v1.ConfigMap
+
+	// First, we try to retrieve from the Cinder CSI-specific config map
+	sourceConfig, err = c.configMapLister.ConfigMaps(util.OpenShiftConfigNamespace).Get("cinder-csi-config")
+	if err != nil {
+		// Failing that, we attempt to retrieve from the cloud provider-specific config map
+		if errors.IsNotFound(err) {
+			sourceConfig, err = c.configMapLister.ConfigMaps(util.OpenShiftConfigNamespace).Get(infra.Spec.CloudConfig.Name)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// TODO: report error after some while?
+					klog.V(2).Infof("Waiting for config map %s from %s", infra.Spec.CloudConfig.Name, util.OpenShiftConfigNamespace)
+					return nil
```

Comment:
So, if we couldn't find configmap and assuming nothing else happend, we won't know we have an error until next 20 mins? Is that what we want?

Reply:
No, it's not what we want and it's something we should improve, but I didn't figure out a better way to do this at the time (https://github.com/openshift/openstack-cinder-csi-driver-operator/pull/78). I'm open to suggestions but while this is not ideal, it is working so I'd like to keep the fix to a separate PR just to avoid swelling this any further.